geopandas/geodataframe.py

```diff
@@ -1701,6 +1702,8 @@ def explode(self, column=None, ignore_index=False, index_parts=None, **kwargs):
         if ignore_index:
             df.reset_index(inplace=True, drop=True)
         elif index_parts:
+            # preserve order because join sometimes sorts index lexicographically
+            df = df.iloc[sorter.sort_index().values]
```

Comment:
I probably need to look closer at the code to understand this :), but: does this change make that now the output of `df.explode()` will always be sorted by the index?

Reply:
No, the original index is preserved and applied on line 1709. This just ensures that if the `join` sorts it, we sort it back here.

Reply:
Ah, yes, I see, it restores the order

Reply:
But, that assumes that `join` _did_ sort / change the order. Is that guaranteed to happen?

According to the docstring (https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.join.html), it should actually preserve the caller's index (*"left: use calling frame’s index (or column if on is specified)"*, but not very clear if that includes the order of the caller's index). 
Knowing pandas .., it might not be always that consistent, and could eg also depend on whether left and right index is equal or not.

Testing out, that's indeed the case:
```python
In [7]: df1 = pd.DataFrame({'a': [1, 2, 3]}, index=[0, 2, 1])

In [8]: df2 = pd.DataFrame({'b': [1, 2, 3]}, index=[0, 2, 1])

In [9]: df1.join(df2)  # <-- preserves index order
Out[9]: 
   a  b
0  1  1
2  2  2
1  3  3

In [10]: df2 = pd.DataFrame({'b': [1, 2, 3, 4]}, index=[0, 2, 1, 1])

In [11]: df1.join(df2)  # <-- index gets sorted
Out[11]: 
   a  b
0  1  1
1  3  3
1  3  4
2  2  2
```

Reply:
So modifying the original bug report example a bit to have only points instead of collections of points (so the explode() basically is a no-op, in which case the join inside `explode()` handles dataframes with equal left and right index):

```
from shapely.geometry import GeometryCollection, Point
import geopandas as gpd

geom = gpd.GeoSeries([Point((0, i)) for i in range(3)])
gdf = gpd.GeoDataFrame(geometry=geom, crs='EPSG:27700')
gdf['id'] = [2, 9, 7]
gdf['x'] = [1, 2, 3]
gdf = gdf.set_index('id')

In [3]: print(gdf)
               geometry  x
id                        
2   POINT (0.000 0.000)  1
9   POINT (0.000 1.000)  2
7   POINT (0.000 2.000)  3

In [4]: gdf.explode(index_parts=True)
Out[4]: 
      x             geometry
id                          
2  0  1  POINT (0.000 0.000)
9  0  3  POINT (0.000 2.000)
7  0  2  POINT (0.000 1.000)
```

So the above is with this PR's branch, and that actually mangled the rows values / index.

Reply:
Good catch. I think I have found a better solution (recent commit). I also added tests to check for these situations.